jinahub/simple/__init__.py

```diff
@@ -0,0 +1,102 @@
+        super().__init__(**kwargs)
+        self._docs = DocumentArrayMemmap(self.workspace + f'/{index_file_name}')
+        self.default_traversal_paths = default_traversal_paths or ['r']
+
+    @requests(on='/index')
+    def index(self, docs: 'DocumentArray', parameters: Dict, **kwargs):
+        """All Documents to the DocumentArray
+        :param docs: the docs to add
+        :param parameters: the parameters dictionary
+        """
+        traversal_path = parameters.get(
+            'traversal_paths', self.default_traversal_paths
+        )
+        flat_docs = docs.traverse_flat(traversal_path)
+        self._docs.extend(flat_docs)
+
+    @requests(on='/search')
+    def search(self, docs: 'DocumentArray', parameters: Dict, **kwargs):
+        """Perform a vector similarity search and retrieve the full Document match
+
+        :param docs: the Documents to search with
+        :param parameters: the parameters for the search"""
+        traversal_path = parameters.get(
+            'traversal_paths', self.default_traversal_paths
+        )
+        flat_docs = docs.traverse_flat(traversal_path)
+        a = np.stack(flat_docs.get_attributes('embedding'))
+        b = np.stack(self._docs.get_attributes('embedding'))
+        q_emb = _ext_A(_norm(a))
+        d_emb = _ext_B(_norm(b))
+        dists = _cosine(q_emb, d_emb)
+        idx, dist = self._get_sorted_top_k(dists, int(parameters['top_k']))
+        for _q, _ids, _dists in zip(flat_docs, idx, dist):
+            for _id, _dist in zip(_ids, _dists):
+                d = Document(self._docs[int(_id)], copy=True)
+                d.scores['cosine'] = 1 - _dist
+                _q.matches.append(d)
+
+    @staticmethod
+    def _get_sorted_top_k(
+        dist: 'np.array', top_k: int
+    ) -> Tuple['np.ndarray', 'np.ndarray']:
+        if top_k >= dist.shape[1]:
+            idx = dist.argsort(axis=1)[:, :top_k]
+            dist = np.take_along_axis(dist, idx, axis=1)
+        else:
+            idx_ps = dist.argpartition(kth=top_k, axis=1)[:, :top_k]
+            dist = np.take_along_axis(dist, idx_ps, axis=1)
+            idx_fs = dist.argsort(axis=1)
+            idx = np.take_along_axis(idx_ps, idx_fs, axis=1)
+            dist = np.take_along_axis(dist, idx_fs, axis=1)
+
+        return idx, dist
+
+
+def _get_ones(x, y):
+    return np.ones((x, y))
+
+
+def _ext_A(A):
+    nA, dim = A.shape
+    A_ext = _get_ones(nA, dim * 3)
+    A_ext[:, dim : 2 * dim] = A
+    A_ext[:, 2 * dim :] = A ** 2
+    return A_ext
+
+
+def _ext_B(B):
+    nB, dim = B.shape
+    B_ext = _get_ones(dim * 3, nB)
+    B_ext[:dim] = (B ** 2).T
+    B_ext[dim : 2 * dim] = -2.0 * B.T
+    del B
+    return B_ext
+
+
+def _euclidean(A_ext, B_ext):
+    sqdist = A_ext.dot(B_ext).clip(min=0)
+    return np.sqrt(sqdist)
+
+
+def _norm(A):
+    return A / np.linalg.norm(A, ord=2, axis=1, keepdims=True)
+
+
+def _cosine(A_norm_ext, B_norm_ext):
+    return A_norm_ext.dot(B_norm_ext).clip(min=0) / 2
```

Comment:
I suggest putting this part into `jina-common` or importing them from `jina`

Reply:
I don't think this is helpful. These functions are too short and not important enough to justify the whole jina_commons install requirement in my opinion.